### PR TITLE
TYP: Overload series/ffill and bfill

### PR DESCRIPTION
**Pull request by pckSF**
- Closes part of MarcoGorelli/pyladieslondon-sprints#29

- Adds `typing.overload` to: `ffill` and ``bfill` from https://github.com/pandas-dev/pandas/blob/master/pandas/core/generic.py

## Setup and Output for `ffill`
```python
# t_ffill.py
import pandas as pd

inplace: bool

reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=False))
reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=True))
reveal_type(pd.DataFrame([1, None, None]).ffill())
reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=inplace))

reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=False))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=True))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=inplace))

reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=False, limit=1))
reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=True, limit=1))
reveal_type(pd.DataFrame([1, None, None]).ffill(limit=1))
reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=inplace, limit=1))

reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=False, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=True, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=inplace, downcast="infer"))

reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=False, limit=1))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=True, limit=1))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, limit=1))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=inplace, limit=1))

reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=False, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=True, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=inplace, downcast="infer"))

reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=False, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=True, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(inplace=inplace, limit=1, downcast="infer"))

reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=False, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=True, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).ffill(axis=0, inplace=inplace, limit=1, downcast="infer"))
```
which results in the following `mypy` output:
```bash
$ mypy t_ffill.py
  t_ffill.py:6: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:7: note: Revealed type is 'None'
  t_ffill.py:8: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:9: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:11: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:12: note: Revealed type is 'None'
  t_ffill.py:13: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:14: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:16: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:17: note: Revealed type is 'None'
  t_ffill.py:18: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:19: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:21: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:22: note: Revealed type is 'None'
  t_ffill.py:23: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:24: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:26: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:27: note: Revealed type is 'None'
  t_ffill.py:28: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:29: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:31: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:32: note: Revealed type is 'None'
  t_ffill.py:33: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:34: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:36: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:37: note: Revealed type is 'None'
  t_ffill.py:38: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:39: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:41: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:42: note: Revealed type is 'None'
  t_ffill.py:43: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:44: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
```

## Setup and Ouput for `bfill`
```python
# t_bfill.py
import pandas as pd

inplace: bool

reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=False))
reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=True))
reveal_type(pd.DataFrame([1, None, None]).bfill())
reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=inplace))

reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=False))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=True))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=inplace))

reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=False, limit=1))
reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=True, limit=1))
reveal_type(pd.DataFrame([1, None, None]).bfill(limit=1))
reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=inplace, limit=1))

reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=False, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=True, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=inplace, downcast="infer"))

reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=False, limit=1))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=True, limit=1))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, limit=1))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=inplace, limit=1))

reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=False, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=True, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=inplace, downcast="infer"))

reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=False, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=True, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(inplace=inplace, limit=1, downcast="infer"))

reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=False, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=True, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, limit=1, downcast="infer"))
reveal_type(pd.DataFrame([1, None, None]).bfill(axis=0, inplace=inplace, limit=1, downcast="infer"))
```
which results in the following `mypy` output:
```bash
$ mypy t_bfill.py
  t_ffill.py:6: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:7: note: Revealed type is 'None'
  t_ffill.py:8: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:9: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:11: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:12: note: Revealed type is 'None'
  t_ffill.py:13: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:14: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:16: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:17: note: Revealed type is 'None'
  t_ffill.py:18: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:19: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:21: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:22: note: Revealed type is 'None'
  t_ffill.py:23: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:24: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:26: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:27: note: Revealed type is 'None'
  t_ffill.py:28: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:29: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:31: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:32: note: Revealed type is 'None'
  t_ffill.py:33: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:34: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:36: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:37: note: Revealed type is 'None'
  t_ffill.py:38: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:39: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
  t_ffill.py:41: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:42: note: Revealed type is 'None'
  t_ffill.py:43: note: Revealed type is 'pandas.core.frame.DataFrame*'
  t_ffill.py:44: note: Revealed type is 'Union[pandas.core.frame.DataFrame*, None]'
```

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -6469,15 +6469,45 @@ def fillna(
...
+        downcast: None | str | dict = ...,
+    ) -> FrameOrSeries | None:
+        ...
+
     @final
     @doc(klass=_shared_doc_kwargs["klass"])
     def ffill(
         self: FrameOrSeries,
-        axis=None,
+        axis: None | str | int = None,
```
Comment (MarcoGorelli) [inline pandas/core/generic.py:6506]: axis has it's own type alias, `Axis` (check the fillna signature)

File: pandas/core/generic.py [hunk 2]
```diff
@@ -6469,15 +6469,45 @@ def fillna(
...
+        limit: None | int = ...,
+        downcast: None | str | dict = ...,
+    ) -> FrameOrSeries:
+        ...
+
+    @overload
+    def ffill(
+        self: FrameOrSeries,
+        axis: None | str | int = ...,
+        inplace: Literal[True] = ...,
```
Comment (MarcoGorelli) [inline pandas/core/generic.py:6486]: this one doesn't need `=...` - else, if you call `.ffill()`, mypy won't know which overload to match

Comment (MarcoGorelli) [reply pandas/core/generic.py:6486]: there also needs to be an extra overload for when `axis` isn't passed:
```python
    @overload
    def ffill(
        self: FrameOrSeries,
        *,
        inplace: Literal[True],

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks Peter, have left some comments

Not too sure about `downcast` but that one can always be left out for now and typed later on

File: pandas/core/generic.py (hunk 1 above)
Comment (pckSF) [reply pandas/core/generic.py]: Done

File: pandas/core/generic.py (hunk 2 above)
Comment (pckSF) [reply pandas/core/generic.py]: Done

File: pandas/core/generic.py [hunk 3]
```diff
@@ -6469,15 +6469,55 @@ def fillna(
...
     def ffill(
         self: FrameOrSeries,
-        axis=None,
+        axis: None | Axis = None,
         inplace: bool_t = False,
-        limit=None,
+        limit: None | int = None,
         downcast=None,
-    ) -> FrameOrSeries | None:
+    ):
```
Comment (MarcoGorelli) [inline pandas/core/generic.py:6520]: I think it should work to keep the return types here, then the function body's return type will be fixed (admittedly this needs fixing in my blog post, will do that now)

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Looks good, thanks! Small comment about the return type in the function itself, other than that looks good to me

File: pandas/core/generic.py (hunk 3 above)
Comment (pckSF) [reply pandas/core/generic.py]: Done

**Review by MarcoGorelli — APPROVED** [review]
Thanks @pckSF !

**Pull request merged**